tower-http/src/compression/body.rs

```diff
@@ -14,124 +14,216 @@ use bytes::{Buf, Bytes};
                 .project()
                 .read
                 .get_pin_mut()
                 .get_pin_mut()
                 .get_pin_mut()
                 .get_pin_mut(),
-            BodyInnerProj::Identity(inner) => inner,
+            BodyInnerProj::Identity { inner } => inner,
+
+            #[cfg(not(feature = "compression-gzip"))]
+            BodyInnerProj::Gzip { inner } => match inner.0 {},
+            #[cfg(not(feature = "compression-deflate"))]
+            BodyInnerProj::Deflate { inner } => match inner.0 {},
+            #[cfg(not(feature = "compression-br"))]
+            BodyInnerProj::Brotli { inner } => match inner.0 {},
         }
     }
 
     /// Consume `self`, returning the inner body
     pub fn into_inner(self) -> B {
-        match self.0 {
+        match self.inner {
             #[cfg(feature = "compression-gzip")]
-            BodyInner::Gzip(inner) => inner
+            BodyInner::Gzip { inner } => inner
                 .read
                 .into_inner()
                 .into_inner()
                 .into_inner()
                 .into_inner(),
             #[cfg(feature = "compression-deflate")]
-            BodyInner::Deflate(inner) => inner
+            BodyInner::Deflate { inner } => inner
                 .read
                 .into_inner()
                 .into_inner()
                 .into_inner()
                 .into_inner(),
             #[cfg(feature = "compression-br")]
-            BodyInner::Brotli(inner) => inner
+            BodyInner::Brotli { inner } => inner
                 .read
                 .into_inner()
                 .into_inner()
                 .into_inner()
                 .into_inner(),
-            BodyInner::Identity(inner) => inner,
+            BodyInner::Identity { inner } => inner,
+
+            #[cfg(not(feature = "compression-gzip"))]
+            BodyInner::Gzip { inner } => match inner.0 {},
+            #[cfg(not(feature = "compression-deflate"))]
+            BodyInner::Deflate { inner } => match inner.0 {},
+            #[cfg(not(feature = "compression-br"))]
+            BodyInner::Brotli { inner } => match inner.0 {},
         }
     }
 }
 
-#[pin_project(project = BodyInnerProj)]
-pub(crate) enum BodyInner<B>
-where
-    B: Body,
-{
+#[cfg(any(
+    not(feature = "compression-gzip"),
+    not(feature = "compression-deflate"),
+    not(feature = "compression-br")
+))]
+pub(crate) enum Never {}
+
+#[cfg(feature = "compression-gzip")]
+type GzipBody<B> = WrapBody<GzipEncoder<B>>;
+#[cfg(not(feature = "compression-gzip"))]
+type GzipBody<B> = (Never, PhantomData<B>);
+
+#[cfg(feature = "compression-deflate")]
+type DeflateBody<B> = WrapBody<ZlibEncoder<B>>;
+#[cfg(not(feature = "compression-deflate"))]
+type DeflateBody<B> = (Never, PhantomData<B>);
+
+#[cfg(feature = "compression-br")]
+type BrotliBody<B> = WrapBody<BrotliEncoder<B>>;
+#[cfg(not(feature = "compression-br"))]
+type BrotliBody<B> = (Never, PhantomData<B>);
```

Comment:
Just checking but are these things necessary because you cannot put `cfg`s inside `pin_project_lite::pin_project!`?

Reply:
Yeah, exactly.